src/App.tsx

```diff
@@ -1,14 +1,85 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, {
+  useEffect, useCallback, useState, useMemo,
+} from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { Todo } from './types/Todo';
+import { getTodos } from './api';
+import { FilterKey } from './types/FilterKey';
+
+function getFilteredTodos(key: FilterKey, query: string, todosToFilter: Todo[]) {
+  let filteredTodos = [];
+
+  switch (key) {
+    case 'all': {
+      filteredTodos = todosToFilter;
+      break;
+    }
+
+    case 'active': {
+      filteredTodos = todosToFilter.filter(todo => !todo.completed);
+      break;
+    }
+
+    case 'completed': {
+      filteredTodos = todosToFilter.filter(todo => todo.completed);
+      break;
+    }
+
+    default:
+      filteredTodos = todosToFilter;
+  }
+
+  filteredTodos = filteredTodos.filter(todo => todo.title
+    .toLowerCase()
+    .includes(query.toLowerCase().trim()));
+
+  return filteredTodos;
+}
 
 export const App: React.FC = () => {
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterKey, setFilterKey] = useState<FilterKey>(FilterKey.All);
+  const [query, setQuery] = useState('');
+
+  const filteredTodos = useMemo(() => {
+    return getFilteredTodos(filterKey, query, todos);
+  }, [filterKey, query, todos]);
+
+  const handleTodoSelection = useCallback((todo: Todo | null) => {
+    setSelectedTodo(todo);
+  }, []);
+
+  const handleFilterKeySelection = useCallback((key: FilterKey) => {
+    setFilterKey(key);
+  }, []);
+
+  const handleQueryChange = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
+    setQuery(event.target.value);
+  }, []);
+
+  const handleQueryDelete = useCallback(() => {
+    setQuery('');
+  }, []);
+
+  useEffect(() => {
+    setIsLoading(true);
+    getTodos()
+      .then(todosFromServer => {
+        setTodos(todosFromServer);
+      })
+      .catch(() => setTodos([]))
```

Comment:
In the `catch()` block, it is necessary to catch an error in order to understand what went wrong.
Fix in all code.